File: python/taflow/metrics/conditional_drawdown_at_risk.py
"""Exact discrete-episode conditional drawdown-at-risk metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import ConditionalDrawdownAtRisk as _Native
from ._input import as_metric_series


class ConditionalDrawdownAtRisk:
    """Compute expected shortfall across discrete drawdown-episode troughs.

    The independent oracle and contract is PerformanceAnalytics 2.1.0 ``CDD`` with
    ``method="discrete"``, geometric compounding, and positive/inverted output.
    Wealth starts at a phantom one. The exact distribution contains the signed
    trough of every contiguous negative or recovered/non-negative drawdown
    segment produced by ``findDrawdowns``. The current unrecovered segment is
    included exactly once. At ``1 - confidence``, the estimator uses R's
    type-7 interpolated quantile and averages all signed troughs at or below the
    boundary, including ties, before returning the positive magnitude.
    ``confidence`` defaults to ``0.95`` and must be strictly between zero and
    one. During warm-up, empty and all-omitted states return ``None``; a
    non-empty path without loss returns ``0.0``.

    This discrete episode contract intentionally differs from Riskfolio-Lib
    7.3.0 ``CDaR_Rel``, which uses the continuous per-observation drawdown path
    and fractional tail weighting. The PerformanceAnalytics source is pinned
    to the CRAN 2.1.0 tarball SHA-256
    ``fc801d39382818cd3a7052326b45d078302aef4d290c85dab83498ed4516d58d``.

    Inputs may be decimal simple returns, log returns, positive equity levels,
    or non-cumulative period P&L. The P&L input method requires positive initial
    equity and Rust performs causal capital conversion. The first equity level
    establishes a baseline and does not increment metric length.
    ``nan_policy`` is ``"omit"`` or ``"raise"``; infinities and simple returns
    below -100% are rejected. Mutating methods are fluent, bulk conversion and
    recurrence execute natively with the GIL released, and repeated unchanged
    ``compute`` calls are O(1). Exact episode order statistics retain O(e)
    memory and a dirty computation is O(e log e), where e is the segment count.
    """

    def __init__(self, confidence: float = 0.95, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(float(confidence), nan_policy)

    def from_returns(
        self, returns: Any, *, column: str | None = None
    ) -> "ConditionalDrawdownAtRisk":
        """Append chronological decimal simple returns and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_log_returns(
        self, log_returns: Any, *, column: str | None = None
    ) -> "ConditionalDrawdownAtRisk":
        """Append chronological log returns and return this metric."""
        self._state.from_log_returns(as_metric_series(log_returns, column=column))
        return self

    def from_equity(
        self, equity: Any, *, column: str | None = None
    ) -> "ConditionalDrawdownAtRisk":
        """Append chronological positive equity levels and return this metric."""
        self._state.from_equity(as_metric_series(equity, column=column))
        return self

    def from_pnl(
        self,
        pnl: Any,
        initial_capital: float,
        *,
        column: str | None = None,
    ) -> "ConditionalDrawdownAtRisk":
        """Append period P&L using required positive initial capital."""
        self._state.from_pnl(
            as_metric_series(pnl, column=column), float(initial_capital)
        )
        return self

    def append(self, value: float) -> "ConditionalDrawdownAtRisk":
        """Append one value in the selected domain and return this metric."""
        self._state.append(float(value))
        return self

    def extend(self, values: Any) -> "ConditionalDrawdownAtRisk":
        """Append a chronological series in the selected domain and return this metric."""
        self._state.extend(as_metric_series(values))
        return self

    @property
    def value(self) -> float | None:
        """Return positive discrete-episode CDaR, or ``None`` while empty."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the cached exact scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "ConditionalDrawdownAtRisk":
        """Clear episodes, preserve configuration/capacity, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable normalized returns processed by Rust."""
        return len(self._state)


__all__ = ["ConditionalDrawdownAtRisk"]

File: crates/taflow-metrics/src/metrics/conditional_drawdown_at_risk.rs
use crate::{
    primitives::DrawdownState, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Exact expected shortfall of discrete drawdown-episode troughs.
#[derive(Debug, Clone)]
pub struct ConditionalDrawdownAtRisk {
    input: MetricInputState,
    drawdown: DrawdownState,
    confidence: f64,
    completed_troughs: Vec<f64>,
    current_trough: Option<f64>,
    current_is_drawdown: bool,
    sorted_troughs: Vec<f64>,
    dirty: bool,
}

impl ConditionalDrawdownAtRisk {
    /// Construct an empty exact discrete-episode CDaR state.
    pub fn new(
        input_kind: MetricInputKind,
        confidence: f64,
        nan_policy: NanPolicy,
    ) -> MetricResult<Self> {
        if !confidence.is_finite() || confidence <= 0.0 || confidence >= 1.0 {
            return Err(MetricError::InvalidParameter {
                name: "confidence",
                value: confidence.to_string(),
                reason: "must be finite and strictly between zero and one",
            });
        }
        if matches!(
            input_kind,
            MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{input_kind:?}"),
                reason: "conditional drawdown at risk requires returns, log returns, equity, or period P&L with initial equity",
            });
        }

        Ok(Self {
            input: MetricInputState::new(input_kind, nan_policy)?,
            drawdown: DrawdownState::new(),
            confidence,
            completed_troughs: Vec::new(),
            current_trough: None,
            current_is_drawdown: false,
            sorted_troughs: Vec::new(),
            dirty: false,
        })
    }

    fn ingest(&mut self, value: f64) -> MetricResult<()> {
        let Some(simple_return) = self.input.append(value)? else {
            return Ok(());
        };
        self.drawdown.append(simple_return)?;
        let drawdown = self
            .drawdown
            .current_drawdown()
            .expect("an appended return always produces a drawdown");
        let is_drawdown = drawdown < 0.0;

        match self.current_trough {
            None => {
                self.current_trough = Some(drawdown);
                self.current_is_drawdown = is_drawdown;
            }
            Some(current) if self.current_is_drawdown == is_drawdown => {
                self.current_trough = Some(current.min(drawdown));
            }
            Some(current) => {
                self.completed_troughs.push(current);
                self.current_trough = Some(drawdown);
                self.current_is_drawdown = is_drawdown;
            }
        }
        self.dirty = true;
        Ok(())
    }

    /// Append one chronological observation and return the current positive CDaR.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        self.ingest(value)?;
        Ok(self.value())
    }

    /// Append a chronological slice and sort the episode distribution at most once.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        self.completed_troughs.reserve(values.len());
        self.sorted_troughs.reserve(values.len().saturating_add(1));
        for &value in values {
            self.ingest(value)?;
        }
        Ok(self.value())
    }

    /// Return exact discrete-episode CDaR, or `None` before a usable return.
    ///
    /// This matches PerformanceAnalytics 2.1.0 `CDD` with `method="discrete"`,
    /// `geometric=TRUE`, and `invert=TRUE`. The signed episode troughs use an R
    /// type-7 quantile at `1 - confidence`; all troughs at or below that
    /// interpolated boundary are averaged, including ties. The active final
    /// episode participates exactly once. Exactness requires O(e) retained
    /// storage for e completed sign episodes; sorting is lazy and only occurs
    /// after input changes.
    pub fn value(&mut self) -> Option<f64> {
        self.current_trough?;
        if self.dirty {
            self.sorted_troughs.clear();
            self.sorted_troughs
                .extend_from_slice(&self.completed_troughs);
            self.sorted_troughs.push(
                self.current_trough
                    .expect("nonempty input has a current drawdown episode"),
            );
            self.sorted_troughs.sort_by(f64::total_cmp);
            self.dirty = false;
        }

        let cutoff = 1.0 - self.confidence;
        let quantile_index = (self.sorted_troughs.len() - 1) as f64 * cutoff;
        let lower = quantile_index.floor() as usize;
        let upper = quantile_index.ceil() as usize;
        let weight = quantile_index - lower as f64;
        let boundary = self.sorted_troughs[lower]
            + (self.sorted_troughs[upper] - self.sorted_troughs[lower]) * weight;

        let mut tail_sum = 0.0;
        let mut tail_count = 0_usize;
        for &trough in &self.sorted_troughs {
            if trough > boundary {
                break;
            }
            tail_sum += trough;
            tail_count += 1;
        }
        Some(-(tail_sum / tail_count as f64))
    }

    /// Return the current exact scalar without replaying prior observations.
    pub fn compute(&mut self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while retaining allocated episode buffers.
    pub fn reset(&mut self) {
        self.input.reset();
        self.drawdown.reset();
        self.completed_troughs.clear();
        self.current_trough = None;
        self.current_is_drawdown = false;
        self.sorted_troughs.clear();
        self.dirty = false;
    }

    /// Return the number of usable normalized returns processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable normalized returns have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
